added management command

jrief opened the pull request:
Can this command be useful?

It sometimes happens, that I have orphaned files in my folder `filer_public`, which do not have any associated meta data. With `./manage.py orphaned_files` we can list them, and with `./manage.py orphaned_files --delete` we can delete them.

yakky commented:
@jrief :+1:

yakky commented:
@jrief I'd just an option to only output filenames instead of "Found orphaned file...", to be able to pipe the output somewhere (I'd use verbosity to control the output form)

jrief commented:
Then the command should be renamed to, say ``filer_check`` and can be used to not only check for orphaned files, but also for missing files. How should we name the options? My proposal is ``--orphans-only``, ``--missing-only`` and ``--delete-orphans``.

yakky commented:
I'd use multiple options:
``--orphans``: scan for orphans
``-missing``: scan for missing
``--delete-orphans``: delete orphans
``--verbose=3``: output verbose string instead of raw filenames

the first two can be combined together to scan both

yakky commented:
@jrief would you have a look at my comment above?

wiesson commented:
Nice idea but it did not work for me because I changed the filer urls in the django settings. WIth `prefix = settings.FILER_STORAGES['public']['main']['UPLOAD_TO_PREFIX']` it worked perfectly.

jrief commented:
@wiesson thanks for the hint, but your approach only works if FILER_STORAGES has been configured explicitly in the project settings. I therefore will wrap it into a try/except-block to fallback to the defaults.

yakky commented:
@jrief any progress on this?

czpython commented on filer/management/commands/orphaned_files.py:
```diff
@@ -0,0 +1,49 @@
+            for filename in files:
+                relfilename = os.path.join(reldir, filename)
+                try:
+                    File.objects.get(file=relfilename)
```
why not `if not File.objects.filter(file=relfilename).exists()`

jrief replied:
fixed

jrief commented:
@yakky thanks for reminding me and sorry for not proceeding on this. I'm stuck with far too many other issues.

benzkji commented:
just curious, does this work if for example S3 storage is configured? Also, does it respect private/public flag on files? Methinks not, as it just uses DefaultStorage() ?

jrief commented:
@yakky 
implemented as you proposed, command is named ``filer_check`` with these options:
* ``--orphans``: scan for orphans
* ``--delete-orphans``: delete orphans
* ``--missing``: scan for missing
* ``--delete-missing``: delete missing
* ``--no-input``/``--noinput`` to skip safety question

@czpython 
thanks for the hint

added unit tests for all command options and entry in CHANGELOG.

sorry for the long delay

jrief commented:
Could someone please explain me, what's wrong with this [failing test](https://travis-ci.org/divio/django-filer/jobs/292451336)?

My local ``isort`` does not complain about my import sorting, and from Travis's output I don't get any feedback.

jrief commented:
@czpython fixed you annotation

jrief commented:
@czpython all annotations have been fixed

jrief commented:
@yakky conflicts have been resolved, but Travis-CI does not want to finish :(

yakky commented:
@jrief travis has completed, but github still says there are conflicts

jrief commented:
@yakky GitHub tells me in big bold letters: **This branch has no conflicts with the base branch**.

Do you really get a different message? Then that must be a GitHub bug.

yakky commented:
@jrief apparently it is
This is what I see :smile: 
![image](https://user-images.githubusercontent.com/714711/40625364-42d835b0-62b2-11e8-8e16-3fe023271614.png)

Applying it manually I have a conflict on the changelog file
Can you try to fetch latest develop (commit id eccd8533de2b76a6793d294b917854eb27dbccae ) and try to merge it locally

czpython commented on docs/management_commands.rst:
```diff
@@ -12,3 +12,33 @@ command.
+
+Invoking::
+
+    ./manage.py filer_check --orphanes
```
typo

jrief replied:
fixed

czpython commented on docs/management_commands.rst:
```diff
@@ -12,3 +12,33 @@ command.
+
+Invoking::
+
+    ./manage.py filer_check --delete-orphanes
```
typo

jrief replied:
fixed

czpython commented on filer/management/commands/filer_check.py:
```diff
@@ -0,0 +1,116 @@
+from filer.settings import DEFAULT_FILER_STORAGES
+
+
+class Command(BaseCommand):
```
management commands shouldn't use mark strings for translation.

jrief replied:
fixed

czpython commented on filer/management/commands/filer_check.py:
```diff
@@ -0,0 +1,116 @@
+                msg = _("\nThis will delete entries from your database. Are you sure you want to do this?\n\n"
+                        "Type 'yes' to continue, or 'no' to cancel: ")
+                if input(msg) != 'yes':
+                    raise CommandError("Aborted: Delete missing file entries from database.")
```
why raise an exception, no is a valid option

jrief replied:
fixed

czpython commented on filer/management/commands/filer_check.py:
```diff
@@ -0,0 +1,116 @@
+                msg = _("\nThis will delete orphaned files from your storage. Are you sure you want to do this?\n\n"
+                        "Type 'yes' to continue, or 'no' to cancel: ")
+                if input(msg) != 'yes':
+                    raise CommandError("Aborted: Delete orphaned files from storage.")
```
same as above.

jrief replied:
fixed

jrief commented:
@yakky merged HEAD of branch develop (eccd8533de2b76a6793d294b917854eb27dbccae) into this branch, with no side effect.

jrief commented:
Unit test job 2196.35 is failing for throttling reasons:

> Access denied | packages.npmjs.org used Cloudflare to restrict access

Please restart the job.

jrief commented:
Ping @czpython 
All request changes have been fixed. Is there still anything missing?

peterfarrell commented:
@yakky is there anything I can do to help get this merged and released?

jrief commented:
Today I had to check the consistency of files in one of my projects. Unfortunately I was quite disappointed that django-filer still lacks this feature, although this pull request is pending and waiting to be merged for one year now.

jrief commented:
Now I rebased this pull request upon the current `development` branch, so it should be ready for the next version.

I really would appreciate, if in version 1.6 this useful feature finally would be available, almost 3 years after starting to work on it.

peterfarrell commented:
@jrief I am at the point where I'm debating forking the project with this feature and releasing it on PyPi. I know I'm not the only person with the need for this functionality.  I do not want to be forced to take that option because it's fragmentation, however the time is approaching quickly.

jrief commented:
Today this pull requests celebrates its 🎂**3rd birthday** 🎂and is still unmerged!

Dear maintainers of django-filer: If you need help with maintenance, please consider to add me to the list of core maintainer of this project.

FinalAngel commented:
@jrief thanks for the PR and changes requested from various reviewers. It gets merged today :)

I also added you to the collaborators, I appreciate the help on Filer 🙏
